### ISSUE TITLE
check bytes of repr(C) enums

### ISSUE THREAD

**Issue by dvc94ch** [rkyv/bytecheck#15]
any way to do it? I guess repr(C) is not defined on enums, although I'd hope it would generate something like this:
```rust
#[repr(C)]
struct {
    discriminant: u8, 
    variant: #[repr(C)] union,
}
```

**Comment by djkoloski** [comment]
`repr(C)` is actually well-defined for enums, and it lays them out as you described. Is the issue that only `repr(Int)` is supported? That has a slightly more efficient layout while still being C-compatible.

**Comment by dvc94ch** [comment]
I'm trying to send archived data structures over a network and use `check_archive_root` to verify the bytes. I didn't realize that I only need to `#[archive_attr(derive(CheckBytes))]` and that that always seems to work.

**Comment by djkoloski** [comment]
Awesome! I'll add an enum example to the book to make it clearer.

**Comment by dvc94ch** [comment]
so reopening. it's still a problem when using `archive(as = "Self")`. although it can be worked around by not using that.

**Comment by djkoloski** [comment]
When using `#[archive(as = "Self")]`, you may need to use `#[derive(CheckBytes)]` on the enum itself instead of in an `#[archive_attr(...)]`. I _think_ that as long as you have the enum set to a compatible repr, it should work. If it doesn't, I can take a closer look at it.

**Comment by dvc94ch** [comment]
Well it's set to repr(C) but it claims it isn't

**Comment by djkoloski** [comment]
Hmm, this might be really dumb but do you have the `#[repr(C)]` before the `#[derive(CheckBytes)]`? The ordering can matter for determining whether some attributes get picked up.

**Comment by dvc94ch** [comment]
Maybe I'm doing something wrong...
```rust
use bytecheck::CheckBytes;

#[derive(CheckBytes)]
#[repr(C)]
pub enum AnEnum {
    Var(u8),
    Other(u16),
}

fn main() {
    println!("Hello, world!");
}
```
```sh
error: archive self enums must be repr(C) or repr(Int)
 --> src/main.rs:4:8
  |
4 | #[repr(C)]
  |        ^

error: aborting due to previous error
```

**Comment by djkoloski** [comment]
Okay, I have located the issue. Thanks for your patience.

That error is outdated and inaccurate. Right now `bytecheck` only has support for deriving validation on `repr(Int)` enums, but `repr(C)` enums could feasibly be supported. That error message will be updated as soon as possible and I'll file an issue for adding `repr(C)` enum support.

**Comment by dvc94ch** [comment]
if that's the case I'm a bit confused why this works:
```rust
use bytecheck::CheckBytes;
use rkyv::Archive;

#[derive(Archive)]
#[archive_attr(derive(CheckBytes))]
#[repr(C)]
pub enum AnEnum {
    Var(u8),
    Other(u16),
}

fn main() {
    println!("Hello, world!");
}
```
is `ArchiveAnEnum` always a `repr(Int)` enum?

**Comment by djkoloski** [comment]
Yes, rkyv will always generate archived enums that are `repr(Int)` even when the original enum is `repr(C)`.

**Comment by djkoloski** [comment]
As an update, I'm no longer planning on adding this myself since I have no need for repr(C) enums. I would not turn down a PR though!
